matching-core/src/main/java/com/graphhopper/matching/MapMatching.java

```
@@ -161,84 +166,201 @@ public MatchResult doWork(List<GPXEntry> gpxList) {
-                            gpxEntry.getLat(), gpxEntry.getLon()) > 2 * measurementErrorSigma
-                    // always include last point
-                    || indexGPX == gpxList.size() - 1) {
-                final List<QueryResult> queryResults = locationIndex.findNClosest(
-                        gpxEntry.lat, gpxEntry.lon,
-                        edgeFilter, measurementErrorSigma);
-                outAllCandidates.addAll(queryResults);
-                final List<GPXExtension> candidates = new ArrayList<>();
-                for (QueryResult candidate : queryResults) {
-                    candidates.add(new GPXExtension(gpxEntry, candidate, indexGPX));
+        int n = filteredGPXEntries.size();
+        assert queriesPerEntry.size() == n;
+        for (int i = 0; i < n; i++) {
+            
+            GPXEntry gpxEntry = filteredGPXEntries.get(i);
+            List<QueryResult> queryResults = queriesPerEntry.get(i);
+
+            /*
+             * If the GPS measurement is mapped to a virtual node then we need to create a
+             * candidate for each direction (see also #51).
+             */
+            List<GPXExtension> candidates = new ArrayList<>();
+            for (QueryResult qr: queryResults) {
+                int closestNode = qr.getClosestNode();
+                if (queryGraph.isVirtualNode(closestNode)) {
+                    // get virtual edges:
+                    List<VirtualEdgeIteratorState> virtualEdges = new ArrayList<>();
+                    EdgeIterator iter = queryGraph.createEdgeExplorer().setBaseNode(closestNode);
+                    while (iter.next()) {
+                        if (queryGraph.isVirtualEdge(iter.getEdge())) {
+                            virtualEdges.add((VirtualEdgeIteratorState)
+                                    queryGraph.getEdgeIteratorState(iter.getEdge(), iter.getAdjNode()));
+                        }
+                    }
+                    assert virtualEdges.size() == 2;
```

Can we throw an exception (IllegalStateException?) here with a detailed message?